File: code/dart/modular_api/lib/src/core/tracing/tracing_middleware.dart
import 'package:dartastic_opentelemetry_api/dartastic_opentelemetry_api.dart';
import 'package:shelf/shelf.dart';

import 'propagation_policy.dart';
import 'server_span.dart';

/// Context key under which the server span is exposed to downstream handlers.
///
/// Reading this is rarely necessary — the span is also made *ambient* for the
/// duration of the request, so `Context.current` finds it and child spans created
/// by `modular_api_postgres` or `modular_api_rest_client` attach automatically.
/// The key exists for code that needs the span object itself.
const String tracingSpanContextKey = 'modular.tracing.span';

/// Context key under which the resolved [PropagationResult] is exposed.
const String propagationResultContextKey = 'modular.tracing.propagation';

/// Creates the middleware that owns the server span.
///
/// **Host-owned, not plugin-owned.** This is installed by the host outside every plugin
/// middleware, which is the whole point: a span created from inside a plugin slot
/// would miss the plugin middleware registered before it, leaving a hole in the
/// waterfall exactly where cold start and early middleware live. ADR-0005
/// decision 4 exists because the first draft of this design got that wrong.
///
/// **Outermost, with `loggingMiddleware` inside it** (runbook D12 as reversed). The
/// logger then runs within this span's active scope, so it reads `trace_id` and
/// `span_id` from the ambient span on every line — including `request completed`,
/// which is emitted from inside the scope. The consequence is that `duration_ms` in
/// the log is slightly *smaller* than the span's duration, which is the right way
/// round: the span is the primary artifact and should cover the most.
///
/// The span is made ambient with `withSpanAsync`, so everything downstream —
/// use cases, database commands, outbound HTTP — can create child spans without
/// anything being threaded through call signatures.
///
/// **This is an adapter, and only an adapter.** Span construction lives in
/// `server_span.dart`, which knows nothing about HTTP or shelf, so the planned gRPC
/// transport arrives as a second adapter rather than a second copy (gate G4).
Middleware tracingMiddleware({
  required APITracer tracer,
  PropagationPolicy policy = const PropagationPolicy(),
  List<String> excludedRoutes = const <String>[],
}) {
  final excluded = Set<String>.from(excludedRoutes);

  return (Handler innerHandler) {
    return (Request request) async {
      final path = request.requestedUri.path;

      // Health, metrics and docs are noise in a trace store. The host derives this
      // list from operationalRoutePaths rather than hardcoding it.
      if (excluded.contains(path)) return innerHandler(request);

      final started = startServerSpan(
        tracer: tracer,
        method: request.method,
        route: path,
        headers: request.headers,
        policy: policy,
      );
      final span = started.span;

      final enriched = request.change(
        context: <String, Object>{
          ...request.context,
          tracingSpanContextKey: span,
          propagationResultContextKey: started.propagation,
        },
      );

      // Ambient for everything downstream: use cases, database commands and
      // outbound HTTP create child spans without anything being threaded through
      // call signatures.
      return tracer.withSpanAsync(span, () async {
        try {
          final response = await innerHandler(enriched);
          completeServerSpan(span, statusCode: response.statusCode);
          return response;
        } catch (error, stackTrace) {
          completeServerSpan(span, error: error, stackTrace: stackTrace);
          rethrow;
        }
      });
    };
  };
}

File: code/ts/modular_api/src/core/tracing/tracingMiddleware.ts
import { context as otelContext, trace, type Tracer } from '@opentelemetry/api';
import type { RequestHandler } from 'express';

import { PropagationPolicy } from './propagationPolicy';
import { completeServerSpan, startServerSpan } from './serverSpan';

/** Property under which the server span is exposed on `res.locals`. */
export const TRACING_SPAN_LOCAL = 'modularTracingSpan';

/** Property under which the resolved propagation result is exposed on `res.locals`. */
export const PROPAGATION_RESULT_LOCAL = 'modularPropagationResult';

export interface TracingMiddlewareOptions {
  readonly tracer: Tracer;
  readonly policy?: PropagationPolicy;
  readonly excludedRoutes?: readonly string[];
}

/**
 * Creates the middleware that owns the server span.
 *
 * **Host-owned, not plugin-owned, and OUTERMOST** — `loggingMiddleware` runs inside it
 * (runbook D12 as reversed), so the logger reads `trace_id` and `span_id` from the
 * ambient span on every line. Being outside every plugin middleware is the other half of
 * the point: a span created from inside a plugin slot would miss the plugin
 * middleware registered before it, leaving a hole in the waterfall exactly where
 * cold start and early middleware live (ADR-0005 decision 4).
 *
 * **Where this genuinely differs from the Dart version, rather than merely reading
 * differently.** Shelf hands a middleware the `Response` its inner handler
 * produced, so the Dart implementation reads the status code and ends the span in a
 * `finally`. Express does not: a middleware calls `next()` and never sees the
 * response. The status code and the end of the request are therefore observed
 * through `res.on('finish')`, with `res.on('close')` covering a client that
 * disconnects before the response completes. A guard makes sure the span ends
 * exactly once whichever fires first.
 *
 * Ambient propagation also differs. Dart's context is zone-based and the API
 * handles it alone; in JavaScript `context.with` needs a context manager, which the
 * application's SDK registers. Without one it degrades to a passthrough: child
 * spans created downstream become roots instead of children, rather than failing.
 * The wiring recipe in the observability guide covers this.
 *
 * **This is an adapter, and only an adapter.** Span construction lives in `serverSpan.ts`, which
 * knows nothing about HTTP or Express, so the planned gRPC transport arrives as a second adapter
 * rather than a second copy (gate G4).
 */
export function tracingMiddleware(options: TracingMiddlewareOptions): RequestHandler {
  const { tracer } = options;
  const policy = options.policy ?? new PropagationPolicy();
  const excluded = new Set(options.excludedRoutes ?? []);

  return (req, res, next) => {
    // Health, metrics and docs are noise in a trace store. The host derives this
    // list from operationalPaths rather than hardcoding it.
    if (excluded.has(req.path)) {
      next();
      return;
    }

    const { span, propagation: resolved } = startServerSpan({
      tracer,
      method: req.method,
      route: req.path,
      headers: req.headers as Record<string, string>,
      policy,
    });

    res.locals[TRACING_SPAN_LOCAL] = span;
    res.locals[PROPAGATION_RESULT_LOCAL] = resolved;

    let ended = false;
    const endSpan = (aborted: boolean) => {
      if (ended) return;
      ended = true;
      completeServerSpan(span, { statusCode: res.statusCode, aborted });
    };

    res.on('finish', () => endSpan(false));
    res.on('close', () => endSpan(!res.writableEnded));

    // Ambient for everything downstream, so a child span created in a use case, a
    // database command or an outbound call attaches by itself.
    otelContext.with(trace.setSpan(resolved.context, span), () => {
      next();
    });
  };
}
